### PR TITLE
[FLINK-26071][table-api-java][table-planner] Now Planner#compilePlan fails if the plan cannot be serialized

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

This PR is a followup of FLINK-25841 and has a couple of quality of life improvements for the new API:

* Now `Planner#compilePlan` fails if the plan cannot be serialized
* Added `CompiledPlan#explain` and use it to implement `toString`, in order to help users access and debug the plan
* Invoking `Planner#loadPlan` will fail on Batch planner, making it consistent with other `*plan` methods
* More Javadoc, in particular around errors

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/BatchPlanner.scala [hunk 1]
```diff
@@ -134,6 +134,11 @@ class BatchPlanner(
     new BatchPlanner(executor, config, moduleManager, functionCatalog, catalogManager)
   }
 
+  override def loadPlan(planReference: PlanReference): CompiledPlanInternal = {
+    throw new UnsupportedOperationException(
+      "The batch planner doesn't support the persisted plan feature.")
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/BatchPlanner.scala:139]: nit: `batch planner` is an internal term, how about `Compiled plans are not supported in batch mode.`

Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/BatchPlanner.scala:139]: also for the exception below

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/StreamPlanner.scala [hunk 2]
```diff
@@ -129,14 +134,48 @@ class StreamPlanner(
...
     val relNodes = modifyOperations.map(translateToRel)
     val optimizedRelNodes = optimize(relNodes)
     val execGraph = translateToExecNodeGraph(optimizedRelNodes)
     cleanupInternalConfigurations()
 
-    new ExecNodeGraphCompiledPlan(createSerdeContext, execGraph)
+    new ExecNodeGraphCompiledPlan(
+      this,
+      JsonSerdeUtil.createObjectWriter(createSerdeContext)
+        .withDefaultPrettyPrinter()
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/StreamPlanner.scala:176]: shall we only pretty print for the `CompiledPlan.printJsonString()`? I guess pretty printing has impact on the plan sizes.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/CompiledPlan.java [hunk 3]
```diff
@@ -79,25 +79,37 @@ default void writeToFile(Path path) throws IOException {
...
     /**
-     * Writes this plan to a file using the JSON representation. This will not overwrite the file if
-     * it's already existing.
+     * Writes this plan to a file using the JSON representation. This operation will be noop if the
+     * file already exists, even if the content is different from this plan.
+     *
+     * @throws IOException if the file cannot be written.
      */
     default void writeToFile(File file) throws IOException {
         writeToFile(file, true);
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/CompiledPlan.java:93]: Is that's the case, then perhaps the method below should have a 3 valued flag: OVERWRITE, IGNORE, FAIL where the default is FAIL?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java [hunk 4]
```diff
@@ -1266,6 +1266,12 @@ void createFunction(
      *
      * <p>This method will parse the input reference and will validate the plan. The returned
      * instance can be executed via {@link #executePlan(CompiledPlan)}.
+     *
+     * @throws IOException if the plan cannot be loaded from the filesystem, or from classpath
+     *     resources.
+     * @throws TableException if the plan is invalid.
+     * @throws UnsupportedOperationException if the {@link EnvironmentSettings} is configured in
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:1273]: I modified it to have both the throws clause in javadoc and the paragraph.

**Review by twalthr — APPROVED** [review]
LGTM

**Comment by twalthr** [comment]
Sorry, I closed this by accident.

**Pull request closed without merging**